app/src/main/java/com/github/polypoly/app/base/game/Game.kt

```diff
@@ -0,0 +1,158 @@
+package com.github.polypoly.app.base.game
+
+import com.github.polypoly.app.base.PastGame
+import com.github.polypoly.app.base.game.location.InGameLocation
+import com.github.polypoly.app.base.game.location.LocalizationLevel
+import com.github.polypoly.app.base.game.rules_and_lobby.GameLobby
+import com.github.polypoly.app.base.game.rules_and_lobby.GameMode
+import com.github.polypoly.app.base.game.rules_and_lobby.GameRules
+import com.github.polypoly.app.base.user.User
+
+/**
+ * Represent the game and the current state of the game
+ * @property admin the [User] who is the admin of the game
+ * @property players the [Player]s of the game
+ * @property rules the rules of the [Game]
+ * @property inGameLocations the [InGameLocation]s of the [Game]
+ * @property currentRound the current round of the [Game]
+ * @property dateBegin the date and time when the [Game] has started in Unix time
+ * (seconds since 1970-01-01T00:00:00Z)
+ */
+class Game private constructor(
+    val admin: User = User(),
+    var players: List<Player> = listOf(),
+    val rules: GameRules = GameRules(),
+    val dateBegin: Long = System.currentTimeMillis(),
+) {
+
+    private val inGameLocations: List<InGameLocation> = rules.gameMap.flatMap { it.locations.map { location ->
+        InGameLocation(
+            location = location,
+            owner = null,
+            level = LocalizationLevel.LEVEL_0,
+        ) } }
+    val currentRound: Int = 1
+
+    /**
+     * Go to the next turn
+     */
+    fun nextTurn() {
+        // TODO update the data with the DB
+        if(isGameFinished()) {
+            val pastGame = endGame()
+            // TODO send the pastGame to the DB
+        }
+    }
+
+    /**
+     * Test if the game is finished
+     * @return true if the game is finished, false otherwise
+     * @throws IllegalStateException if the game mode is RICHEST_PLAYER and maxRound is null
+     */
+    private fun isGameFinished(): Boolean {
+        return when(rules.gameMode) {
+            GameMode.LAST_STANDING -> {
+                players.filter { !it.hasLose() }.size <= 1
+            }
+            GameMode.RICHEST_PLAYER -> {
+                if(rules.maxRound == null)
+                    throw IllegalStateException("maxRound can't be null in RICHEST_PLAYER game mode")
+                currentRound > rules.maxRound
+            }
+        }
+    }
+
+    /**
+     * Return the ranking of the players
+     * @return a [Map] of the [Player]s' [User] id and their rank
+     */
+    fun ranking(): Map<Long, Int> {
+        val playersSorted = players.sortedDescending()
+        val map = playersSorted.mapIndexed { index, player -> player.user.id to index+1 }.toMap().toMutableMap()
+        for(i in 1 until (players.size)) {
+            if(playersSorted[i].compareTo(playersSorted[i-1]) == 0) {
+                map[playersSorted[i-1].user.id]?.let{ map[playersSorted[i].user.id] = it }
+            }
+        }
```

Comment:
This could be refactor using two iterator variable `currentPlayer` and `previousPlayer` instead of using several times `playersSorted[i]` and `playersSorted[i-1]`
Moreover, I believe a simple affectation in the map would be clearer than the `let` + lambda

Reply:
thank you, fix done